### DISCUSSION TITLE
QuestPDF. Dynamic columns and rows

### DISCUSSION THREAD

**Discussion by Craigoman79** [QuestPDF/QuestPDF#872] [Q&A]
I want to have PDF content to reflect incoming and outgoing messages. I am using columns with 2 rows to reflect left and right indentation based on whether the message is outgoing or incoming. This seems to work fine using Constant and Relative Items. I ideally would want each row to adjust based on the message content so smaller messages take up less space while longer ones take up more and they should be able to overlap. 

It currently looks like this:
![image](https://github.com/QuestPDF/QuestPDF/assets/160397690/1797732f-e5c5-41de-a77f-be9ac1581f45)

Is there a better way of achieving this? This is what my codes currently looks like:  

```
Document.Create(document =>
{
    document.Page(page =>
    {
        page.Size(PageSizes.A4);
        page.Margin(1, Unit.Centimetre);

        page.DefaultTextStyle(x => x.FontSize(10));

        page.Header().Element(Header);

        page.Content()
            .PaddingVertical(25)
            .Column(column =>
            {
                column.Spacing(5);

                int totalMessages = 10;

                foreach (var i in Enumerable.Range(1, totalMessages))
                {
                    column.Item().Row(row =>
                    {
                        if (i % 2 == 0) // Outgoing messages (left-aligned)
                        {

                            row.ConstantItem(200); // Adjust the size here to fit your layout needs for right alignment
                            row.RelativeItem()
                                .Background(Colors.Green.Lighten2)
                                .Border(1, Unit.Point)
                                .Padding(1, Unit.Millimetre)
                                .Column(innerColumn =>
                                {
                                    innerColumn.Item().Text($"Sender {i}").FontSize(10).Bold();
                                    string text = "Sample Outgoing Message text for message " + i;
                                    innerColumn.Item().Text(text).FontSize(10);
                                    innerColumn.Item().Text("30/04/2024 14:20PM").FontSize(8).FontColor(Colors.Grey.Darken1).AlignRight();
                                });
                            
                        }
                        else // Incoming messages (right-aligned)
                        {
                            
                            row.RelativeItem()
                                .Background(Colors.Grey.Lighten2)
                                .Border(1, Unit.Point)
                                .Padding(1, Unit.Millimetre)
                                .Column(innerColumn =>
                                {
                                    innerColumn.Item().Text($"Sender {i}").FontSize(10).Bold();
                                    string text = "Sample Incoming Message text for message " + i;
                                    innerColumn.Item().Text(text).FontSize(10);
                                    innerColumn.Item().Text(Placeholders.DateTime()).FontSize(8).FontColor(Colors.Grey.Darken1).AlignRight();
                                });

                            row.ConstantItem(200); // Adjust the size here to fit your layout needs for right alignment
                        }
                    });
                }
            });

        page.Footer().Element(Footer);
    });
}).ShowInPreviewer();

```

**Comment by MarcinZiabek** [comment]
This is my attempt, a good starting point for you.

```csharp
.Column(column => 
{
    column.Spacing(20);

    column.Item()
        .AlignLeft()
        .Element(MessageBox)
        .Text(Placeholders.Label());

    column.Item()
        .AlignRight()
        .Element(MessageBox)
        .Text(Placeholders.Paragraph());
    
    column.Item()
        .AlignLeft()
        .Element(MessageBox)
        .Text(Placeholders.Sentence());
    
    column.Item()
        .AlignRight()
        .Element(MessageBox)
        .Height(50)
        .Image(Placeholders.Image(50, 20))
        .FitArea();
    
    IContainer MessageBox(IContainer container) => container.MaxWidth(300).Background(Colors.Grey.Lighten4).Padding(5);
});
```

That produces:

<img width="523" alt="image" src="https://github.com/QuestPDF/QuestPDF/assets/9263853/5a0ccc76-b702-4f11-bc73-e397d832bf48">

**Reply by Craigoman79** [reply]
That looks great. Can I add a few bits of information to each bubble? Sender, Image, Message, Timestamp? Is there any option to round the corners of the bubbles?

**Reply by MarcinZiabek** [reply]
The rounded corners are a bit harder at this moment. You can do it by integrating the library with SkiaSharp. There is an example in official documentation. I am planning to introduce this feature natively.

In terms of content, instead of text/images, you can include any layout structures, for example tables with more sophisticated content. But everything should work.

**Comment by Craigoman79** [comment]
Ok makes sense. I have modified my code to look like this. 2 things. 
1. How can I scale to image to be smaller and not take up all the space - so force it to 50x20 and resize the column accordingly?
2. Ensure that Messages are not split over pages.

![image](https://github.com/QuestPDF/QuestPDF/assets/160397690/e979a96f-3041-4bcd-9c70-3fa2eae52d54)

`Document.Create(document =>
{
    document.Page(page =>
    {
        page.Size(PageSizes.A4);
        page.Margin(1, Unit.Centimetre);

        page.DefaultTextStyle(x => x.FontSize(10));

        page.Header().Element(Header);

        page.Content()
            .PaddingVertical(25)
            .Column(column =>
            {
                column.Spacing(5);

                int totalMessages = 10;

                foreach (var i in Enumerable.Range(1, totalMessages))
                {

                    column.Item()
                        .AlignLeft()
                        .Background(Colors.Grey.Lighten4)
                        .MaxWidth(400)
                        .Padding(5)
                        .Column(innerColumn =>
                        {
                            innerColumn.Item().Text($"Sender {i}").FontSize(10).Bold();
                            string text = "Sample Outgoing Message text for message " + i;
                            innerColumn.Item().Text(text).FontSize(10);
                            innerColumn.Item().Text("30/04/2024 14:20PM").FontSize(8).FontColor(Colors.Grey.Darken1).AlignRight();
                        });

                    column.Item()
                        .AlignRight()
                        .Background(Colors.Green.Lighten4)
                        .MaxWidth(400)
                        .Padding(5)
                        .Column(innerColumn =>
                        {
                            innerColumn.Item().Text($"Sender {i}").FontSize(10).Bold();
                            string text = Placeholders.LoremIpsum();
                            innerColumn.Item().Text(text).FontSize(10);
                            innerColumn.Item().Text("30/04/2024 14:20PM").FontSize(8).FontColor(Colors.Grey.Darken1).AlignRight();
                        });

                    column.Item()
                        .AlignRight()
                        .Background(Colors.Green.Lighten4)
                        .MaxWidth(400)
                        .Padding(5)
                        .Column(innerColumn =>
                        {
                            innerColumn.Item().Text($"Sender {i}").FontSize(10).Bold();
                            innerColumn.Item().Image(Placeholders.Image(50,20)).FitArea();
                            string text = Placeholders.Sentence();
                            innerColumn.Item().Text(text).FontSize(10);
                            innerColumn.Item().Text("30/04/2024 14:20PM").FontSize(8).FontColor(Colors.Grey.Darken1).AlignRight();
                        });

                    column.Item()
                        .AlignLeft()
                        .Background(Colors.Grey.Lighten4)
                        .MaxWidth(400)
                        .Padding(5)
                        .Column(innerColumn =>
                        {
                            innerColumn.Item().Text($"Sender {i}").FontSize(10).Bold();
                            string text = Placeholders.Sentence();
                            innerColumn.Item().Text(text).FontSize(10);
                            innerColumn.Item().Text("30/04/2024 14:20PM").FontSize(8).FontColor(Colors.Grey.Darken1).AlignRight();
                        });

                }
            });

        page.Footer().Element(Footer);
    });
}).ShowInPreviewer();`

**Comment by Craigoman79** [comment]
I resolved the splitting over pages using .ShowEntire(). Just not sure why the Image is not forced to the width and height I want rather than taking up the full space.

**Reply by MarcinZiabek** [reply]
You need to constrain the image element to physical size:

```csharp
innerColumn.Item().Width(100).Image(Placeholders.Image(50,20)).FitArea();
```

**Reply by Craigoman79** [reply]
Great stuff. makes sense. Thanks for the quick reply

**Comment by Craigoman79** [comment]
I have decided to rather keep the sender and timestamp on the first row. I have amended my code to reflect this but the timestamp does not right align in the row. Any suggestions? 

![image](https://github.com/QuestPDF/QuestPDF/assets/160397690/0005f157-e440-4f6a-8fd2-5df4c6bba798)


```
 void AddMessageContent(ColumnDescriptor innerColumn, MessageDetails message, string tempImageDirectory)
 {
     // Row for sender and timestamp
     innerColumn.Item().Row(row =>
     {
         row.AutoItem().Text(message.Sender).FontSize(8).Bold();
         row.AutoItem().AlignRight().Text(message.Timestamp.ToString("g")).FontSize(6).FontColor(Colors.Grey.Darken1);
     });
```

**Reply by Craigoman79** [reply]
any suggestion here?

**Reply by MarcinZiabek** [reply]
When using only `AutoItem`s, the `Row` element will attempt to take as little space as possible. You can force it to take entire width by putting a `RelativeItem`:

```csharp
innerColumn.Item().Row(row =>
{
   row.AutoItem().Text(message.Sender).FontSize(8).Bold();
   row.RelativeItem();
   row.AutoItem().Text(message.Timestamp.ToString("g")).FontSize(6).FontColor(Colors.Grey.Darken1);
});
```

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

We hope the replies above were helpful in building your chat-style layout.

Should you encounter any further problems, please don't hesitate to open a new issue.
Thank you!
